receiver/datadogreceiver/translator.go

```diff
@@ -50,60 +49,75 @@ func toTraces(payload *pb.TracerPayload, req *http.Request) ptrace.Traces {
+		semconv.AttributeTelemetrySDKName:      "Datadog",
+		semconv.AttributeTelemetrySDKVersion:   payload.TracerVersion,
+	} {
+		if v != "" {
+			sharedAttributes.PutStr(k, v)
+		}
+	}
+
+	for k, v := range payload.Tags {
+		if k = translateDataDogKeyToOtel(k); v != "" {
+			sharedAttributes.PutStr(k, v)
+		}
+	}
+
+	upsertHeadersAttributes(req, sharedAttributes)
+
+	// Creating a map of service spans to slices
+	// since the expectation is that `service.name`
+	// is added as a resource attribute in most systems
+	// now instead of being a span level attribute.
+	groupByService := make(map[string]ptrace.SpanSlice)
 
 	for _, trace := range traces {
-		ils := resSpans.ScopeSpans().AppendEmpty()
-		ils.Scope().SetName("Datadog-" + req.Header.Get("Datadog-Meta-Lang"))
-		ils.Scope().SetVersion(req.Header.Get("Datadog-Meta-Tracer-Version"))
-		spans := ptrace.NewSpanSlice()
-		spans.EnsureCapacity(len(trace))
 		for _, span := range trace {
-			newSpan := spans.AppendEmpty()
+			slice, exist := groupByService[span.Service]
+			if !exist {
+				slice = ptrace.NewSpanSlice()
+				groupByService[span.Service] = slice
+			}
+			newSpan := slice.AppendEmpty()
 
 			newSpan.SetTraceID(uInt64ToTraceID(0, span.TraceID))
 			newSpan.SetSpanID(uInt64ToSpanID(span.SpanID))
 			newSpan.SetStartTimestamp(pcommon.Timestamp(span.Start))
 			newSpan.SetEndTimestamp(pcommon.Timestamp(span.Start + span.Duration))
 			newSpan.SetParentSpanID(uInt64ToSpanID(span.ParentID))
-			newSpan.SetName(span.Resource)
+			newSpan.SetName(span.Name)
+			newSpan.Status().SetCode(ptrace.StatusCodeOk)
 
 			if span.Error > 0 {
 				newSpan.Status().SetCode(ptrace.StatusCodeError)
-			} else {
-				newSpan.Status().SetCode(ptrace.StatusCodeOk)
 			}
 
-			attrs := newSpan.Attributes()
-			attrs.EnsureCapacity(len(span.GetMeta()) + 1)
-			attrs.PutStr(semconv.AttributeServiceName, span.Service)
 			for k, v := range span.GetMeta() {
-				k = translateDataDogKeyToOtel(k)
-				if len(k) > 0 {
-					attrs.PutStr(k, v)
+				if k = translateDataDogKeyToOtel(k); len(k) > 0 {
+					newSpan.Attributes().PutStr(k, v)
 				}
 			}
 
-			if span.Meta["span.kind"] != "" {
-				switch span.Meta["span.kind"] {
-				case "server":
-					newSpan.SetKind(ptrace.SpanKindServer)
-				case "client":
-					newSpan.SetKind(ptrace.SpanKindClient)
-				case "producer":
-					newSpan.SetKind(ptrace.SpanKindProducer)
-				case "consumer":
-					newSpan.SetKind(ptrace.SpanKindConsumer)
-				case "internal":
-					newSpan.SetKind(ptrace.SpanKindInternal)
-				default:
-					newSpan.SetKind(ptrace.SpanKindUnspecified)
-				}
-			} else {
+			switch span.Meta[datadogSpanKindKey] {
+			case "server":
+				newSpan.SetKind(ptrace.SpanKindServer)
+			case "client":
+				newSpan.SetKind(ptrace.SpanKindClient)
+			case "producer":
+				newSpan.SetKind(ptrace.SpanKindProducer)
+			case "consumer":
+				newSpan.SetKind(ptrace.SpanKindConsumer)
+			case "internal":
+				newSpan.SetKind(ptrace.SpanKindInternal)
+			default:
```

Comment:
Why did you get rid of default?

Reply:
I got rid of the `if` statement around it since it could be merged in with the switch case statement for performance gains (it is very slight).